MK94.Assert.Mocking/Mocker.cs

```diff
@@ -0,0 +1,137 @@
+﻿using Castle.DynamicProxy;
+using MK94.Assert.Chain;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MK94.Assert.Mocking
+{
+    /// <summary>
+    /// A class to generate mocks that expect calls from <see cref="DiskAsserter.Operations"/> and setup returns via <see cref="DiskAsserter.Matches{T}(string, T)"/>
+    /// </summary>
+    public class Mocker : IInterceptor
+    {
+        private static readonly ProxyGenerator proxyGenerator = new ProxyGenerator();
+
+        private readonly DiskAsserter diskAsserter;
+
+        private AsyncLocal<Queue<AssertOperation>> operations = new AsyncLocal<Queue<AssertOperation>>();
+        private AsyncLocal<int> count = new AsyncLocal<int>();
+
+        public Mocker(DiskAsserter diskAsserter)
+        {
+            this.diskAsserter = diskAsserter;
+        }
+
+        /// <summary>
+        /// Creates a mock of <typeparamref name="T"/>
+        /// </summary>
+        /// <typeparam name="T">The class or interface to mock</typeparam>
+        /// <param name="actual">The implementation to use when running in write mode. <br />
+        /// The calls and results of this implementation will be record for future re-runs.</param>
+        /// <returns>A mocked instance of <typeparamref name="T"/></returns>
+        public T Of<T>(Func<T> actual)
+            where T : class
+        {
+            if (diskAsserter.WriteMode && actual != default)
+                return proxyGenerator.CreateInterfaceProxyWithTarget(actual(), this);
+
+            return proxyGenerator.CreateInterfaceProxyWithoutTarget<T>(this);
+        }
+
+
+        /// <summary>
+        /// Creates a mock of <typeparamref name="T"/>. Useful for a builder pattern.
+        /// </summary>
+        /// <typeparam name="T">The class or interface to mock</typeparam>
+        /// <param name="actual">The implementation to use when running in write mode. <br />
+        /// The calls and results of this implementation will be record for future re-runs.</param>
+        /// <returns>A mocked instance of <typeparamref name="T"/></returns>
+        /// <param name="mocked">The result mock object</param>
+        /// 
+        /// <example>
+            /// <code>
+                /// 
+                /// DiskAssert.Default
+                ///     .Of&lt;T1&gt;(() => new ImplementationOfT(), out var mocked1)
```

Comment:
Is this correct?

Reply:
yea shows an example in rider, but not visual studio for some reason
<img width="507" alt="image" src="https://user-images.githubusercontent.com/5530511/154340650-a3589bb9-b292-407c-bcca-24b6f4ab00bb.png">
tbh i'm not sure if i'm using it correctly